feature(Add stock movement in the patient financial Status)

lomamech opened the pull request:
- Implement API for get the stock movement related to the Patient
- (test) add integration test for get patients/:uuid/stock/movements
- Sanitaze code
closes #4141

lomamech committed:
feature(Add stock movement in the patient financial Status)
- Implement API for get the stock movement related to the Patient
- (test) add integration test for get patients/:uuid/stock/movements
- Sanitaze code
closes #4141

lomamech committed:
test: reset result test value

lomamech committed:
test: reset result test value

lomamech committed:
test: reset result test value in find movements by lot name

lomamech committed:
test: reset result test value in find by lots reasons for distribution to patient

lomamech committed:
test: reset result test value in find movements by reference

lomamech committed:
test: reset result test value in find entry movements

lomamech committed:
test: reset result test value in find entry movements

lomamech committed:
(test): reset result for many test in stock movement

lomamech committed:
test: fix data value in find entry movements

lomamech committed:
test: fix value for find movements by reference

jniles commented on client/src/js/components/bhPatientFinancialActivity.js:
```diff
@@ -59,6 +59,16 @@ function PatientFinancialActivityCtrl(Patients, moment, Session, Constants) {
         $ctrl.loading = false;
       });
+
+    Patients.getStockMovements($ctrl.patientUuid)
```
@lomamech, could you combine the `Patient.getFinancialActivity()` and `Patient.getStockMovements()` call with a `Promise.all()` and end it with the `.finally()`?

Like this:
```js
Promise.all([
 Patients.getFinancialActivity($ctrl.patientUuid),
 Patients.getStockMovements($ctrl.patientUuid),
]).then([financialData, stockData] => {
  // ... etc ...
})
.finally(() =>{
  $ctrl.loading = false;
});
```

The reason for doing this is that we don't want to set `$ctrl.loading = false` too early.

jniles commented on client/src/js/components/bhPatientFinancialActivity.js:
```diff
@@ -59,6 +59,16 @@ function PatientFinancialActivityCtrl(Patients, moment, Session, Constants) {
+    Patients.getStockMovements($ctrl.patientUuid)
+      .then(data => {
+        $ctrl.dataMovement = data;
+        $ctrl.noStockMovement = data.length === 0;
```
If there is no financial activity (but there is stock activity), this component will be covered up with a message.  Here is an example:
![image](https://user-images.githubusercontent.com/896472/78898904-7cf97000-7a6c-11ea-9b39-00ecb25bfe63.png)

There is stock data here, but no financial data.  Could you change [this line](https://github.com/IMA-WorldHealth/bhima/pull/4269/files#diff-8422e147a296cedc27c69a1710e1cd2aL21) from:
```html
<bh-panel-overlay ng-if="$ctrl.noFinancialHistory">
```
to
```html
<bh-panel-overlay ng-if="$ctrl.noFinancialHistory && $ctrl.noStockMovement">
```

jniles commented on client/src/modules/patients/patients.service.js:
```diff
@@ -345,6 +346,19 @@ function PatientService(
+   */
+  function getStockMovements(uuid) {
+    const path = 'patients/:uuid/stock/movements';
+    return service.$http.get(path.replace(':uuid', uuid))
```
Just FYI - you can also do this: 
```js
 const path = `patients/${uuid}/stock/movements`;
```
and skip the `.replace()` part.

jniles commented on client/src/modules/patients/record/bh-patient-financial-activity.html:
```diff
@@ -79,6 +79,33 @@ <h5 ng-show="!$ctrl.status.isInGoodStanding" class="text-danger">
+              </tr>
+            </thead>
+            <tbody>
+              <tr ng-repeat="dataMovement in $ctrl.dataMovement">
```
We should probably limit this to the most recent ... 5 items?  Or something like that.  We don't want the grid to go on forever!

I would just implement this on the client side for the moment.

jniles commented on server/controllers/medical/patients/index.js:
```diff
@@ -678,3 +679,28 @@ function getDebtorBalance(req, res, next) {
+ * @description
+ * returns the stock Movements to the patient.
+ */
+function getStockMovements(req, res, next) {
```
Beautiful.  Does what it needs to do, and nothing else :+1:

jniles commented on client/src/modules/patients/record/bh-patient-financial-activity.html:
```diff
@@ -79,6 +79,33 @@ <h5 ng-show="!$ctrl.status.isInGoodStanding" class="text-danger">
         </dl>
+
+        <hr/>
+        <h5 class="text-success" style="margin-top:0;">
```
Can you remove the `text-success` here?  I'm not sure it makes sense to use here.

![image](https://user-images.githubusercontent.com/896472/78900392-b0d59500-7a6e-11ea-9506-0b5526069656.png)

jniles reviewed (commented):
@lomamech, this is great!  Well done!  I used it with Vanga's database, and it immediately gave me a lot more power to ask questions of the data.

There are a few corner cases that I've highlighted in the comments.  Could you take a look at those?  After that, it LGTM.  Great job!

lomamech committed:
feature(Stock Movement in Patient financial status)
- Combine the getFinancialActivity and getStockMovements
- Display stock movement when they are no financial history
- Sanitaze code in patients service
- Set limit this to the most recent 5 items in client side

jniles reviewed (approved):
Looks like everything as passed!  Excellent implementation.

Bors r+
